File: Number-of-Islands.py
1import queue
2class Solution:
3    def bfs(self, grid: List[List[str]],row: int,col: int,R: int,C: int) -> None:
4        diff = [0,1,0,-1,0]
5        grid[row][col] = 0
6        q = queue.Queue()
7        q.put([row,col])
8        while(not q.empty()):
9            val = q.get()
10            currentrow,currentcol = val[0],val[1]
11            for di in range(4):
12                adjR,adjC = currentrow + diff[di], currentcol + diff[di+1]
13                if(adjR>=0 and adjR<R and adjC>=0 and adjC<C and grid[adjR][adjC] == "1"):
14                    grid[adjR][adjC] = "0"
15                    q.put([adjR,adjC])
16
17    def numIslands(self, grid: List[List[str]]) -> int:
18        R,C = len(grid), len(grid[0])
19        area = 0
20        for row in range(R):
21            for col in range(C):
22                if(grid[row][col] == "1"):
23                    self.bfs(grid,row,col,R,C)
24                    area+=1
25        return area

File: Number-of-Islands.java
1class Solution {
2    static int[] diff = {0,1,0,-1,0};
3    public void bfs(char[][] grid, int row, int col, int R, int C){
4        grid[row][col] = '0';
5        Queue<int[]> q = new LinkedList<>();
6        q.add(new int[]{row,col});
7
8        while(!q.isEmpty()){
9            int[] val = q.poll();
10            int currentrow = val[0], currentcol = val[1];
11            for(int di = 0 ; di < 4 ; di++){
12                int adjR = currentrow + diff[di], adjC = currentcol + diff[di+1];
13                if(adjR>=0 && adjR<R && adjC>=0 && adjC<C && grid[adjR][adjC] == '1'){
14                    grid[adjR][adjC] = '0';
15                    q.add(new int[]{adjR,adjC});
16                }
17            }
18        }
19    }
20    public int numIslands(char[][] grid) {
21        int R = grid.length, C = grid[0].length;
22        int area = 0;
23        for(int row = 0 ; row < R ; row++){
24            for(int col = 0 ; col < C ; col++){
25                if(grid[row][col] == '1'){
26                    area++;
27                    bfs(grid,row,col,R,C);
28                }
29            }
30        }
31        return area;
32    }
33}